packages/dcos-integration-test/extra/test_networking.py

```diff
@@ -1,37 +1,170 @@
+    proxy_app['constraints'] = [['hostname', 'CLUSTER', host2]]
+    proxy_app['mem'] = 0
+    proxy_app['cpu'] = 0
+    if r.proxynet == 'USER':
+        proxy_app['ipAddress'] = {'networkName': 'dcos'}
+    if r.proxynet == 'HOST':
+        proxy_app['cmd'] = '/opt/mesosphere/bin/dcos-shell python '\
+                           '/opt/mesosphere/active/dcos-integration-test/util/python_test_server.py $PORT0'
+        proxy_app['container']['docker']['portMappings'][0]['hostPort'] = 0
+        proxy_app['container']['docker']['portMappings'][0]['containerPort'] = 0
+
+    returned_uuid = None
+
+    with contextlib.ExitStack() as stack:
+        stack.enter_context(dcos_api_session.marathon.deploy_and_cleanup(origin_app, timeout=timeout))
+        sp = stack.enter_context(dcos_api_session.marathon.deploy_and_cleanup(proxy_app, timeout=timeout))
+        cmd = '/opt/mesosphere/bin/curl -s -f -m 5 http://{}/test_uuid'.format(r.vipaddr)
+        returned_uuid = ensure_routable(cmd, sp)
+        log.debug('returned_uuid is: {}'.format(returned_uuid))
+    assert returned_uuid is not None
+    assert returned_uuid['test_uuid'] == app_uuid
+    r.log('PASSED')
+
+
+@pytest.fixture
+def reduce_logging():
+    start_log_level = logging.getLogger('test_util.marathon').getEffectiveLevel()
+    # gotta go up to warning to mute it as its currently at info
+    logging.getLogger('test_util.marathon').setLevel(logging.WARNING)
+    yield
+    logging.getLogger('test_util.marathon').setLevel(start_log_level)
+
+
+@pytest.mark.skipif(not lb_enabled(), reason='Load Balancer disabled')
+def test_vip(dcos_api_session, reduce_logging):
+    """Test every permutation of VIP
+    """
+    addrs = [['1.1.1.{}:{}', '1.1.1.{}:{}'],
+             ['/namedvip{}:{}', 'namedvip{}.marathon.l4lb.thisdcos.directory:{}']]
+    samehosts = [True, False]
+    vipnets = ['USER', 'BRIDGE', 'HOST']
+    proxynets = ['USER', 'BRIDGE', 'HOST']
+    permutations = [[vi, va, sh, vn, pn]
+                    for [vi, va] in addrs
+                    for sh in samehosts
+                    for vn in vipnets
+                    for pn in proxynets]
+    tests = [VipTest(i, vi, va, sh, vn, pn) for i, [vi, va, sh, vn, pn] in enumerate(permutations)]
+    threads = []
+    # deque is thread safe
+    failed_tests = deque(tests)
+    passed_tests = deque()
+    skipped_tests = deque()
+    for r in tests:
+        if not r.samehost and len(dcos_api_session.slaves) == 1:
+            failed_tests.remove(r)
+            skipped_tests.append(r)
+            continue
+
+        def run(test):
+            vip_test(dcos_api_session, test)
+            failed_tests.remove(test)
+            passed_tests.append(test)
+
+        threads.append(threading.Thread(target=run, args=([r])))
+    groups = [threads[i:i + maxthreads] for i in range(0, len(threads), maxthreads)]
+    for gs in groups:
+        for t in gs:
+            t.start()
+        for t in gs:
+            t.join()
```

Comment:
For next time :-) https://docs.python.org/dev/library/concurrent.futures.html#threadpoolexecutor